lib/memo_wise.rb

```diff
@@ -195,9 +195,39 @@ def #{method_name}(#{MemoWise::InternalAPI.args_str(method)})
               end
             end
           HEREDOC
-        # MemoWise::InternalAPI::MULTIPLE_REQUIRED, MemoWise::InternalAPI::SPLAT,
-        # MemoWise::InternalAPI::DOUBLE_SPLAT, MemoWise::InternalAPI::SPLAT_AND_DOUBLE_SPLAT
-        else
+        when MemoWise::InternalAPI::MULTIPLE_REQUIRED
+          # When we have multiple required params, we store the results in a
+          # deeply nested hash, like:
+          # { method_name: { arg1 => { arg2 => { arg3 => memoized_value } } } }
+          last_index = method.parameters.size
+          layers = method.parameters.map.with_index(1) do |(_, name), index|
+            prev_hash = "_memo_wise_hash#{index - 1 if index > 1}"
+            fallback = if index == last_index
+                         "#{original_memo_wised_name}(#{MemoWise::InternalAPI.call_str(method)})"
+                       else
+                         "{}"
+                       end
+            "_memo_wise_hash#{index} = #{prev_hash}.fetch(#{name}) { #{prev_hash}[#{name}] = #{fallback} }"
+          end
+          klass.module_eval <<~HEREDOC, __FILE__, __LINE__ + 1
+            def #{method_name}(#{MemoWise::InternalAPI.args_str(method)})
+              _memo_wise_hash = (@_memo_wise[:#{method_name}] ||= {})
+              #{layers.join("\n  ")}
+            end
+          HEREDOC
+        when MemoWise::InternalAPI::SPLAT_AND_DOUBLE_SPLAT
+          klass.module_eval <<~HEREDOC, __FILE__, __LINE__ + 1
+            def #{method_name}(*args, **kwargs)
+              _memo_wise_hash = (@_memo_wise[:#{method_name}] ||= {})
+              _memo_wise_kwargs_hash = _memo_wise_hash.fetch(args) do
+                _memo_wise_hash[args] = {}
+              end
+              _memo_wise_kwargs_hash.fetch(kwargs) do
+                _memo_wise_kwargs_hash[kwargs] = #{original_memo_wised_name}(#{MemoWise::InternalAPI.call_str(method)})
+              end
+            end
+          HEREDOC
+        else # MemoWise::InternalAPI::SPLAT, MemoWise::InternalAPI::DOUBLE_SPLAT
```

Comment:
Didn't there used to be `MemoWise::InternalAPI::SPLAT_AND_DOUBLE_SPLAT` as well?

Reply:
Oops, didn't see it! nevermind